spotlight-client/src/stores/UserStore.ts

```diff
@@ -0,0 +1,116 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+// =============================================================================
+
+import createAuth0Client, { Auth0ClientOptions } from "@auth0/auth0-spa-js";
+import { makeAutoObservable, runInAction } from "mobx";
+import qs from "qs";
+import { ERROR_MESSAGES } from "../constants";
+import RootStore from "./RootStore";
+
+type ConstructorProps = {
+  authSettings?: Auth0ClientOptions;
+  isAuthRequired: boolean;
+  rootStore?: RootStore;
+};
+
+/**
+ * Reactive wrapper around Auth0 client.
+ * If auth is disabled for this environment, all users will be immediately authorized.
+ * Otherwise, call `authorize` to retrieve credentials or start login flow.
+ *
+ * @example
+ *
+ * ```js
+ * const store = new UserStore({ isAuthRequired: true, authSettings: { domain, client_id, redirect_uri } });
+ * if (!store.isAuthorized) {
+ *   await store.authorize();
+ *   // this may trigger a redirect to the Auth0 login domain;
+ *   // if we're still here and user has successfully logged in,
+ *   // store.isAuthorized should now be true.
+ * }
+ * ```
+ */
+export default class UserStore {
+  authSettings?: Auth0ClientOptions;
+
+  awaitingVerification: boolean;
+
+  isAuthorized: boolean;
+
+  isLoading: boolean;
+
+  rootStore?: RootStore;
+
+  constructor({ authSettings, isAuthRequired, rootStore }: ConstructorProps) {
+    makeAutoObservable(this);
+
+    this.authSettings = authSettings;
+    this.rootStore = rootStore;
+
+    this.awaitingVerification = false;
+    if (!isAuthRequired) {
+      this.isAuthorized = true;
+      this.isLoading = false;
+    } else {
+      this.isAuthorized = false;
+      this.isLoading = true;
+    }
+  }
+
+  /**
+   * If user already has a valid Auth0 credential, this method will retrieve it
+   * and update class properties accordingly. If not, user will be redirected
+   * to the Auth0 login domain for fresh authentication.
+   * Returns an Error if Auth0 configuration is not present.
+   */
+  async authorize(): Promise<Error | void> {
+    if (!this.authSettings) {
+      // returning rather than throwing for better type safety
+      return new Error(ERROR_MESSAGES.auth0Configuration);
+    }
+    const auth0 = await createAuth0Client(this.authSettings);
+
+    const urlQuery = qs.parse(window.location.search, {
+      ignoreQueryPrefix: true,
+    });
+    if (urlQuery.code && urlQuery.state) {
+      await auth0.handleRedirectCallback();
+      // auth0 params are single-use, must be removed or they'll cause errors
+      const newUrl = new URL(window.location.href);
+      delete urlQuery.code;
+      delete urlQuery.state;
+      newUrl.search = qs.stringify(urlQuery, { addQueryPrefix: true });
+      window.history.replaceState({}, document.title, newUrl.href);
```

Comment:
I'm wondering if this is more complicated than it needs to be. In pulse-dashboard we use:
```
const { appState } = await auth0FromHook.handleRedirectCallback();
window.history.replaceState(
    {},
    document.title,
    appState && appState.targetUrl
      ? appState.targetUrl
      : window.location.pathname
);
```

Are there other requirements here that require the manual manipulation of the `urlQuery`?

Reply:
no, you are right ... the original thinking here is that we are likely to be using query params for some application state and I didn't want to clobber them. But then I got sidetracked and never got around to replacing this with something better!